### PR TITLE
Fix handling of bounded variable-length char arrays in BINARY2 format

### LINKED ISSUE #8737: votable binary parsing appears to fail on null columns
Example:

```
>>> import requests
>>> url = "http://vizier.u-strasbg.fr/viz-bin/votable/-b64?-source=V/127A/mash1&-out.max=20&-out.meta=hu2D"
>>> result = requests.get(url)
>>> with open('test.vot', 'w') as fh:
...       fh.write(result.text)
>>> from astropy.table import Table
>>> tbl = Table.read('test.vot')
>>> print(tbl)
```

results in:
```
<Table masked=True length=18>
n_PNG      PNG           Name     ...  ObsDate    img   AssocData
                                  ...     d
bytes1   bytes12       bytes13    ...   int32    bytes5   bytes4
------ ------------ ------------- ... ---------- ------ ---------
     T G209.1-08.2  PHR0615-0025  ...  808464672  fitsP      G227
     . 3-12.0 PHR06 33-1808 @X��� ... 1718187123  TG212      .2-0
     4 .7 PHR0633-0 135 @X�z�G�� ... 1346843185  4.2-0      2.4
     P HR0645-0217  @YP������a�P ...  858666541  06.8       PHR0
     6 46-1235 @Yf� �,_��)2��n\B ...  858667296  PHR06      48-0
     7 19 @Y��z�G�
R��,_�B
 ... 1346916912  650+0      013
     @ Y������?�'�}         '�|B� ...  892415280  257 @       Y��

 �Q���
��r               ...  892411968  Y�,_�      ���#
     �       Q�
�Bp               ... 1506109904  6��)      Z+<M
     ^          oC;               ...  744472613  �g���       �A�
                 A�             E ...  801727228    �BD        B<
                    �1012 fitsLG2 ...  155354624
                    3 fitsTG217.2 ...    4338176
     E            a tsTG214.6+02. ...          0                R
     8 1015 fitsTG2 27.2-03.4 PHR ...          1      E      1016
       fitsTG222.9- 01.1 PHR0705- ...    2452667  1017       fits
     P G237.9-07.2  FP0711-2531   ...  825241912   fits      TG22
     6 .4-01.3 PHR0 711-1238 @Z�� ...  543582580  sLG22      5.2+
```

If you drop the 'CS' column (which is a null column), the table parses correctly, so it looks like the astropy votable is incorrectly handling the null column.

cc @gilleslandais

### LINKED ISSUE #18099: Is the VOTable parser incorrectly handling variable-length char arrays by stripping the asterisk from the arraysize attribute?
### Description

In the process of developing a VOTable binary2 serializer, we've encountered an issue where queries that return a VOTable result with fields that have a variable char datatype with a bounded arraysize seem to fail, in ether a silent way empty table or with an exception.

As an example if we try to run a query that returns just the access_format of our ObsCore table which has the following field definition:

```
<FIELD name="access_format" datatype="char" arraysize="128*" ucd="meta.code.mime" utype="Access.format">
    <DESCRIPTION>Content format of the dataset</DESCRIPTION>
</FIELD>
```

This results in an empty table. We've confirmed with other clients like TopCat and those do in fact display the table rows.

Upon some quick investigation into how astropy is parsing BINARY2 tables, what seems to be happening is that the binary2 converting process handles bounded variable-length char fields by stripping the asterisk in arraysize="128*" and treating it as a fixed-length field, causing it to attempt reading 128 bytes.

Specifically, the string fields end up using their Char Converter, where the * is stripped from the arraysize and then astropy switches the binary parsing method to `_binparse_fixed`.

This method reads fixed-length arrays by reading a fixed number of bytes (specified by self.arraysize) which seems incorrect for this field type.

I think what should be used in this case is the `_binparse_var` method, where we get the length from the 4-byte length prefix and read exactly as many bytes as needed.
This aligns with what we see in the IVOA specification and our understanding that 128* means "up-to 128 length".

The relevant sections are:

https://github.com/astropy/astropy/blob/9fce0d46c5e1807d7e1030c3cb0b1a9c0a359dd9/astropy/io/votable/converters.py#L299
https://github.com/astropy/astropy/blob/9fce0d46c5e1807d7e1030c3cb0b1a9c0a359dd9/astropy/io/votable/converters.py#L326
https://github.com/astropy/astropy/blob/9fce0d46c5e1807d7e1030c3cb0b1a9c0a359dd9/astropy/io/votable/converters.py#L376C9-L376C24
https://github.com/astropy/astropy/blob/9fce0d46c5e1807d7e1030c3cb0b1a9c0a359dd9/astropy/io/votable/converters.py#L376
https://github.com/astropy/astropy/blob/9fce0d46c5e1807d7e1030c3cb0b1a9c0a359dd9/astropy/io/votable/converters.py#L372

Perhaps someone could confirm if this interpretation of the issue is correct or if we've misunderstood how this should work?
If so I could perhaps put in a quick PR to address this.

### Expected behavior

The expectation would be that the the parser reads the 4-byte length prefix (via _parse_length), then reading exactly that many bytes and decoding them.

### How to Reproduce

1. Parse a VOTable file where fields are variable-length char arrays (e.g. 128*)

```python
  from astropy.io.votable import parse as votable_parse
  votable = votable_parse('result.xml')
```

A sample VOTable that reproduces the issue:

```
<?xml version="1.0" encoding="UTF-8"?>
<VOTABLE xmlns="http://www.ivoa.net/xml/VOTable/v1.3" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" version="1.4">
  <RESOURCE type="results">
    <INFO name="QUERY_STATUS" value="OK" />
    <INFO name="QUERY" value="SELECT TOP 1 access_format FROM ivoa.ObsCore" />
    <TABLE>
      <FIELD name="access_format" datatype="char" arraysize="128*" ucd="meta.code.mime" utype="Access.format">
        <DESCRIPTION>Content format of the dataset</DESCRIPTION>
      </FIELD>
      <DATA>
      <BINARY2>
        <STREAM encoding='base64'>
AAAAACphcHBsaWNhdGlvbi94LXZvdGFibGUreG1sO2NvbnRlbnQ9ZGF0YWxpbms=
        </STREAM>
      </BINARY2>
    </DATA>
</TABLE>
  </RESOURCE>
</VOTABLE>

```

### Versions

```
platform
--------
platform.platform() = 'Linux-6.8.0-58-generic-x86_64-with-glibc2.35'
platform.version() = '#60~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Fri Mar 28 16:09:21 UTC 2'
platform.python_version() = '3.12.7'

packages
--------
astropy              7.2.dev28+g5c2a9f1b22
numpy                2.2.0
scipy                --
matplotlib           --
pandas               --
pyerfa               2.0.1.5
```

### PR DESCRIPTION
**Pull request by stvoutsin**
### Description

This pull request is to address issue #18099 with VOTable BINARY2 serialization where fields with bounded variable length char arrays (e.g., arraysize="128*") were being handled as fixed length. 

Specifically, the parser was removing the asterisk from the arraysize attribute and treating these fields as fixed-length arrays, which caused them to be incorrectly encoded/decoded in BINARY2 format.

The issue resulted in either empty tables or exceptions when parsing VOTable results containing these field types.

### Changes

- Modified the Char converter class's `__init__` method to detect bounded variable length arrays (e.g. arraysize="128*") and use the appropriate var parsing methods
- Preserved the original arraysize format with the asterisk to ensure it's correctly written back during serialization
- Added bounds checking in variable length array handlers to properly validate against the max lengths
- Added unit tests to verify correct handling of bounded variable length char arrays
- Changed existing unit tests to ensure these field type metadata is properly propagated through the roundtrips. 

### Technical Details

The issue was caused by this code in the Char converter's `__init__` method:
`field.arraysize = field.arraysize.removesuffix("*")`

This was removing the asterisk and therefore losing the information that this should be handled as a variable length field with an upper bound, which later in the `__init__` method lead to the fixed length binary parsing methods being used. The changes should address this by recognizing that the field is a variable length field with an upper bound and assigning the converted the appropriate binary parsing methods. The changes also address the round-tripping of this information between astropy/votable tables.

### Testing
Added following new tests:

- `test_binary2_single_bounded_char_array`:     Test that variable length char arrays with a max length arraysize="128*") are read correctly in BINARY2 format.
- `test_binary2_bounded_variable_length_char`: Tests round-trip serialization of bounded variable length char arrays across all formats
- `test_binary2_bounded_variable_length_char_edge_cases`: Tests edge cases
- `test_binary2_char_fields_vizier_data`: Test parsing a VOTable from a Vizier query which includes bounded variable-length char arrays (e.g., arraysize="7*"). Related issue: https://github.com/astropy/astropy/issues/8737

This fix is particularly important for interoperability with IVOA services that commonly use this field type in their response formats.

### Related Issues
Fixes #18099 
Fixes #8737

### Notes: 
Let me know if I've misinterpreted the issue or if there are better ways to address it and I'd be happy to make any changes accordingly.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by stvoutsin** [comment]
Looks like the test data files I added are not properly being resolved in the GH workflow. I've set this PR as a draft while I sort out what went wrong.

File: astropy/io/votable/tests/test_table.py [hunk 1]
```diff
@@ -518,6 +518,174 @@ def test_validate_output_valid():
     assert validate_out is True
 
 
+def test_binary2_single_bounded_char_array():
+    """
+    Test that variable length char arrays with a max length
+    (arraysize="128*") are read correctly in BINARY2 format.
+    """
+    votable = parse(get_pkg_data_filename("data/binary2_variable_length_char.xml"))
```
Comment (pllim) [inline astropy/io/votable/tests/test_table.py:526]: It is choking here. I do not see this file added in this PR. Are you sure this is the correct filename?

Comment (stvoutsin) [reply astropy/io/votable/tests/test_table.py:526]: Yea I must have not added that data file in the original commit, I've added it now thanks

**Comment by pllim** [comment]
You can ignore the "dev version" allowed failure job. I restarted RTD.

**Review by ManonMarchand — COMMENTED** [review]
Thanks for the PR. I checked in the branch, and it works.

My only comment is that the tests could have an `unicodeChar` field instead of one of the `char` ones.

**Comment by stvoutsin** [comment]
> Thanks for the PR. I checked in the branch, and it works.
> 
> My only comment is that the tests could have an `unicodeChar` field instead of one of the `char` ones.

Thanks for the comment and for checking the branch @ManonMarchand .I've added a couple more unit tests for checking UnicodeChars & for validating handling of bounds.

**Review by ManonMarchand — APPROVED** [review]

File: astropy/io/votable/converters.py [hunk 2]
```diff
@@ -389,6 +404,11 @@ def _binoutput_var(self, value, mask):
                 value = value.encode("ascii")
             except ValueError:
                 vo_raise(E24, (value, self.field_name))
+
+        if self.arraysize != "*" and len(value) > self.arraysize:
+            vo_warn(W46, ("char", self.arraysize), None, None)
+            value = value[: self.arraysize]
```
Comment (bsipocz) [inline astropy/io/votable/converters.py:410]: is this a reasonable behaviour when `len(value) > self.arraysize`, should this maybe raise an exception rather than just a warning?

File: astropy/io/votable/converters.py [hunk 3]
```diff
@@ -424,14 +444,24 @@ def __init__(self, field, config=None, pos=None):
             self.binoutput = self._binoutput_var
             self.arraysize = "*"
         else:
-            field.arraysize = field.arraysize.removesuffix("*")
+            is_variable = field.arraysize.endswith("*")
+
+            # Get numeric part without modifying field.arraysize
+            numeric_part = field.arraysize.removesuffix("*")
```
Comment (bsipocz) [inline astropy/io/votable/converters.py:450]: We could uniformize the comments here with between L326:L330 as currently they are complimenter for the same code content.

File: astropy/io/votable/converters.py [hunk 4]
```diff
@@ -459,8 +493,15 @@ def _binparse_fixed(self, read):
...
         if mask or value is None or value == "":
             return _zero_int
+
         encoded = value.encode("utf_16_be")
-        return self._write_length(len(encoded) / 2) + encoded
+
+        if self.arraysize != "*" and len(value) > self.arraysize:
+            vo_warn(W46, ("unicodeChar", self.arraysize), None, None)
+            value = value[: self.arraysize]
+            encoded = value.encode("utf_16_be")
```
Comment (bsipocz) [inline astropy/io/votable/converters.py:502]: remove the same line from L497 and deindent this one

**Review by bsipocz — APPROVED** [review]
I only have minor comments, take or leave them as you wish.

File: astropy/io/votable/converters.py (hunk 3 above)
Comment (stvoutsin) [reply astropy/io/votable/converters.py]: Yes good point, should be done in new commit

File: astropy/io/votable/converters.py (hunk 2 above)
Comment (stvoutsin) [reply astropy/io/votable/converters.py:409]: That's an interesting point, are you suggesting we raise an exception in the output methods or the parse methods as well? I've gone with raising warning to keep consistent with the default Char parse method and generally in favour of being more forgiving for users, but I'm happy to also raise an exception if you think this is more appropriate here. If so do you want me to do the same for the parsing method?

Comment (bsipocz) [reply astropy/io/votable/converters.py:409]: Yes, but neither the Char parser no here in e.g. L385 has any value loss, while in this method you chop the return at arraysize. 
Maybe keep it a warning but don't chop?

Comment (stvoutsin) [reply astropy/io/votable/converters.py:409]: Thinking about this I think for consistency it may be best to truncate in the parser and the output method.
Providing a warning without truncating has potential to cause unexpected compatibility issues later on for users, for example if a user pre-allocates fixed buffers based on declared length. Also it feels wrong to generate a Table that will not conform with the schema.

If we didn't want to truncate, I'd also consider increasing the arraysize along with a warning, but that may also lead to unexpected issues.

Any thoughts on truncating both at parse and output? Otherwise I can also just not chop, the potential for the error occurring through some round-trip where on the other end they assume data that conforms with the bound is probably rare.

Comment (bsipocz) [reply astropy/io/votable/converters.py:409]: I would lean towards not truncating in here, so it will be consistent (if wrong) with the rest of the methods, and thus can be easily be backported. Then this PR can go in now without further discussion.

But then opening a bug report issue where to ask for input, and possibly changing the behaviour, either by truncating or changing the schema. The former feels more correct to me, but I would like to see the input from others, too. Then potentially a fix/refactor; that is consistent with the Char methods, too, can go in the next feature release as a breaking change.

Comment (stvoutsin) [reply astropy/io/votable/converters.py]: Ok sounds good, I've removed the truncation for the output method. I've also added a few more test methods since I noticed codecov was complaining about the coverage. Let me know if you think anything else needs to be changed.

Comment (bsipocz) [reply astropy/io/votable/converters.py]: Thanks!

Not sure why codecov is still complaining, I would suggest to ignore it for now and go ahead with the merge once CI is all green. 

Would you mind opening an issue outlining the concerns you mentioned in https://github.com/astropy/astropy/pull/18105#discussion_r2100985240 ?

**Review by bsipocz — APPROVED** [review]
Thanks!

File: astropy/io/votable/converters.py (hunk 2 above)
Comment (stvoutsin) [reply astropy/io/votable/converters.py]: By the way I noticed the CI failed during the link check here:

`(api/astropy.io.votable.validator.make_validation_report: line   23) broken    http://www.starlink.ac.uk/stilts - HTTPConnectionPool(host='www.starlink.ac.uk', port=80): Max retries exceeded with url: /stilts (Caused by ConnectTimeoutError(<urllib3.connection.HTTPConnection object at 0x7f33be2d7350>, 'Connection to www.starlink.ac.uk timed out. (connect timeout=180)'))`

Is this a known issue? I noticed that it is using an older URL for the starlink docs:
https://github.com/astropy/astropy/blob/16a94b4d4f9f5070115dac7f52a3a1bdfbdca2cb/docs/conf.py#L612

This should be the new URL I think:
https://www.star.bristol.ac.uk/mbt/stilts/

Comment (bsipocz) [reply astropy/io/votable/converters.py]: Unrelated, so ignore it or open a separate PR.

(actually I'm not sure why we require the extra CIs for this PR, but at this point I'm just waiting for all the jobs on the exotic arc to finish)

File: astropy/io/votable/converters.py [hunk 5]
```diff
@@ -459,8 +490,14 @@ def _binparse_fixed(self, read):
     def _binoutput_var(self, value, mask):
         if mask or value is None or value == "":
             return _zero_int
+
+        if self.arraysize != "*" and len(value) > self.arraysize:
+            vo_warn(W46, ("unicodeChar", self.arraysize), None, None)
+            value = value[: self.arraysize]
```
Comment (bsipocz) [inline astropy/io/votable/converters.py:496]: Ahh, review mistake, I didn't spot that this line still truncate and  is now inconsistent with the same method in `Char`

Comment (pllim) [reply astropy/io/votable/converters.py:496]: Is a follow-up PR planned? Thanks!

Comment (stvoutsin) [reply astropy/io/votable/converters.py:496]: Oh I didn't notice that inconsistency, my bad!
